app/src/heartbeat/heartbeat_storage_desktop.cc

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "app/src/heartbeat/heartbeat_storage_desktop.h"
+
+#include <fstream>
+
+#include "app/logged_heartbeats_generated.h"
+#include "app/src/filesystem.h"
+#include "app/src/include/firebase/internal/mutex.h"
+
+namespace firebase {
+namespace heartbeat {
+
+using com::google::firebase::cpp::heartbeat::GetLoggedHeartbeats;
+using com::google::firebase::cpp::heartbeat::LoggedHeartbeats;
+using com::google::firebase::cpp::heartbeat::VerifyLoggedHeartbeatsBuffer;
+
+// Using an anonymous namespace for file mutex and filename
+namespace {
+
+const char kHeartbeatDir[] = "firebase-heartbeat";
+const char kHeartbeatFilename[] = "HEARTBEAT_STORAGE";
+
+// Returns the mutex that protects accesses to the storage file.
+Mutex& FileMutex() {
+  static Mutex* mutex_ = new Mutex(Mutex::kModeNonRecursive);
+  return *mutex_;
+}
+
+std::string GetFilename(std::string& error) {
+  std::string app_dir =
+      AppDataDir(kHeartbeatDir, /*should_create=*/true, &error);
+  if (app_dir.empty()) {
+    return "";
+  }
+
+  // TODO(almostmatt): does this need to be different between windows/mac/linux?
+  return app_dir + "/" + kHeartbeatFilename;
```

Comment:
When I searched online it seemed like fstream might have internal logic to convert / and \ for windows. I haven't verified this yet, though in theory some of the added unit tests would confirm this if run on windows.

Reply:
Ok, I have confirmed that this does indeed work.
I wrote a simple script that uses fstream to write to "C:/Users/username/Documents/filename1" and also use fstream to write to "C:\\Users\\username\\Documents\\filename2", and ran the script on a windows device. Both files were created.

Reply:
It looks like the windows implementation of AppDataDir also uses forward slashes. I think that if we were to start using backslashes for windows implementation we would want to be consistent and update the existing AppDataDir function as well, at which point I think it would make sense to be a separate change from this heartbeats implementation.
I'm happy to do so if you would like.